### ISSUE TITLE
IE 11 only - Using ngx-formly and accordion-group to show it in the list.

### ISSUE THREAD

**Issue by bobbytechie** [dhilt/ngx-ui-scroll#202]
I can't replicate the exact steps in the stackblitz as it requires ngx-formly and ngx-accordion.

As you know, IE takes time to render the form elements in the list.

My approach will be similar to this [Demo](https://stackblitz.com/edit/ngx-ui-scroll-1-7-0-angular-10-fjmyhe)

In IE Browser only, after adding like 5 to 15 items at a time and scrolling down at some point of time getting below error.

![image](https://user-images.githubusercontent.com/45222207/90044857-df4cd080-dceb-11ea-88d1-f62e7a1d5b69.png)

@dhilt

**Comment by dhilt** [comment]
@bobbytechie In your sample when you do insert, you are relying on the data from the App component's level:

```js
      this.datasource.adapter.insert({
        after: ({ data }) => data === this.data[size - 1],
        items
      })
```

which seems suspicious, because current Scroller Buffer may not contain all the App data at the moment. Instead, I would suggest to use Adapter.append with `eof` set to true, if you need to add items to the end of the dataset. Also, a good practice would be to `clip` extra items right after append/insert:

```js
this.datasource.adapter.append({ items, bof: true }).then(() => 
  this.datasource.adapter.clip()
)
```

At last, you should protect Adapter insert/append calls by waiting for the relaxation of the Scroller. Adapter calls should follow in series. Otherwise, if you run insert/append too fast or render is too slow, there may be some internal interference breaking the process.

```js
adapter.relax().then(() =>
  adapter.append({ items, bof: true }).then(() => 
    adapter.clip()
  )
)
```

or 

```js
await adapter.relax()
await adapter.append({ items, bof: true })
adapter.clip()
```

Let me know, if it helps.

PS Both ngx-formly and ngx-accordion could be added to stackblitz demo, as it already has ngx-ui-scroll, which is an external dependency.

**Comment by bobbytechie** [comment]
@dhilt Thank you for in detailed explanation. For me, most of the time getting when scrolling. Seems the scroll is fast and the items are taking time to render. Is there anything that we could do with success callback?

[Demo Link with Formly and Accordion](https://stackblitz.com/edit/angular-g6mifv)

Height will change based on accordion open/close, which I am using Adapter.check() in it. 

It is working good in Chrome but failing in **IE Browser** mostly when am scrolling and used to click on accordion open/close.

Thanks!

**Comment by dhilt** [comment]
@bobbytechie Thanks for the demo, though it was not easy to run it under IE. After some investigation I can state that my previous comment contains the answer: fast append and slow render cause race conditions and you need to provide append calls in series.

```ts
doAppend(items: any[]) {
  if (!this.datasource.adapter.isLoading) { // immediate append is the Scroller relaxes
    this.datasource.adapter.append({ items });
    return;
  }
  this.datasource.adapter.relax(() => this.doAppend(items)); // try to append after the Scroller finishes its work
}
```

Use this method instead of `this.datasource.adapter.append` call inside `ngDoCheck`.

Also, I would think about grouping optimization. For example:

 - user makes his first "doAppend" call (1)
 - append process runs
 - user makes another two "doAppend" calls (2 and 3) when the first one is not finished
 - currently it will try to run both appends after the first one is done, and (2) will go immediately while (3) will be postponed until (2) is done

Now imagine we have 10 "doAppend" concurrent calls. Each time it will try to run ALL - 1 appends, adding new items one by one. The better approach would be to combine postponed appends, something like the following:

```ts
itemsToAppend = [];

doAppend(items?: any[]) {
  if (!this.datasource.adapter.isLoading) {
    this.datasource.adapter.append({ items: items || this.itemsToAppend });
    this.itemsToAppend = [];
    return;
  }
  if (!this.itemsToAppend.length) {
    this.datasource.adapter.relax(() => this.doAppend());
  }
  this.itemsToAppend = [ ...this.itemsToAppend, ...items ];
}
```

I didn't try this code, but hope you'll get the idea.

**Comment by dhilt** [comment]
And clip, run `Adapter.clip` after append. Virtualization can suffer without it (in case append is the only way the Scroller gets new items).

**Comment by bobbytechie** [comment]
Still seeing same error @dhilt in IE and I didn't understand where to add `Adapter.clip`

We have three scenarios here
1) Create
2) Update
3) Accordion open/close

After creating all the investments here then also need to show all on the update.

I am having issues in all three cases like appending and also while showing all during an update.

sometimes when I am trying to close accordions one by one at a time.

I agree with you on `fast append and slow render` but need a slow solution for the above three cases.

I have updated demo with 
`
  // model = {
  //   investments: [{}]
  // };
`
`model = investments;`

**In the create(while appending new with above fix).**
![image](https://user-images.githubusercontent.com/45222207/90228174-648ecd00-de33-11ea-8a41-2f46e1550762.png)

**In the update(Loading already added data).**
![image](https://user-images.githubusercontent.com/45222207/90228194-6b1d4480-de33-11ea-9e9e-5aa9f403ed4d.png)

**Comment by dhilt** [comment]
Here's the new issue with IE11 (https://github.com/dhilt/ngx-ui-scroll/issues/205) and I'm planning to switch to it as it seems critical, more clear and potentially could have an impact on what you are trying to deal with.

I like how the new `doAppend` method works, especially merging optimization. I was incorrect, I failed with running your demo in IE, but I run it in Edge and I saw "HTML element is not defined" error (relevant to your screenshots) on append. So we have fixed append case for Edge. And I believe IE also should pass multiple append scenario after it.

Also, I can confirm that accordion open/close case has the same problem. I see it in Edge. It happens when I close some accordion leading to new items fetching and then quickly open it before new items are rendered. My guess is that you may also protect this kind of operations with `adapter.isLoading`/`adapter.relax` logic. The main principle is the same: Adapter actions must follow in sequence, not in parallel. In this scenario you need to protect `Adapter.check` calls. Build your own transactions bus as it was done for the append scenario. This might be a little bit overkill, but it will guarantee a consistent actions flow.

_______

Another much simpler solution might be just to block the UI when `Adapter.isLoading$` carries `true`. Just disable controls and forget about Adapter race conditions.

**Comment by bobbytechie** [comment]
@dhilt Please do let me know once your fix is ready for Scroll. Thank You.

In between, I will try for append and Adapter.check.

**Comment by bobbytechie** [comment]
@dhilt Any update?

**Comment by dhilt** [comment]
@bobbytechie  I believe the ball's in your court. We have investigated the cause of your issue, it is about race conditions between Adapter calls. We have developed one of the possible fixes for one of your scenarios. My guess is that it should be enough to continue. I would agree that the lib could handle such race conditions by itself, but it will be very expensive feature and I think it will bring up more questions than answers. Anyway, I would not wait for the fix on the lib's end for now, though the case seems important, it can be handled on the App side by one of the following options:

 - manage the Adapter actions bus to provide synchronous Adapter methods calls;
 - block the UI when the Scroller is busy (like `[disabled]="datasource.adapter.isLoading ? true : null"`).

Btw, ngx-ui-scroll v1.7.1 is released, it affects IE behaviour but only for the entire window case.  Does not relate to your issue.

**Comment by bobbytechie** [comment]
@dhilt I tried it by disabling but initial loading also not loading all the items properly instead showing HTML element is not defined in console.

I agree with keeping wait while adding new items or while open/close accordions but what about the callbacks in the datasource get method?
